dev-docs/source/streaming.rst

```diff
@@ -0,0 +1,185 @@
+
+Python offers a generator pattern to read data while it's being retrieved.
+Django also has ``QuerySet.iterator()`` to process data one record at a time
+and a ``StreamingHttpResponse`` class that allows the WSGI server to read
+the response body from a generator/iterator.
+
+Unfortunately, each step inside the Django REST Framework pipeline breaks the streaming behavior.
+Whenever code reads all all incoming data upfront, the whole chain of streaming records breaks.
+Hence, all steps were altered to work with generators:
+
+.. graphviz::
+
+   digraph foo {
+        rankdir=LR
+        node [shape=cds height=0.7]
+        edge [style=invis minlen=1]
+
+        collect [label="QuerySet.iterator()"]
+        collect2 [label="ChunkedQuerySetIterator"]
+        paginate [label="paginate without list()"]
+        serialize [label="serializer.data\nReturnGenerator"]
+        paginate2 [label="wrap in page"]
+        render [label="render single objects"]
+        footer [label="write footer"]
+        response [label="start StreamingResponse"]
+
+        collect -> paginate -> serialize -> paginate2 -> response -> render -> footer
+        collect2 -> paginate
+        {rank=same; collect; collect2}
+        collect -> collect2 [xlabel="or"]
+   }
+
+The following steps are taken:
+
+* Serializers read using ``QuerySet.iterator()`` whenever possible.
+* Serializers return a ``ReturnGenerator`` instead of a ``ReturnList``.
+* The paginator delegates most rendering to the output format; it only adds the basic structure.
+* The next/previous links are determined *after* rendering all main objects.
+* Our custom ``HALJSONRenderer`` and ``GeoJSONRenderer`` classes support generators.
+* The rendering classes perform ``json.dumps()`` calls on single records.
+* The ``Response`` class is replaced by a ``StreamingResponse`` class.
+
+Chunked JSON Rendering
+----------------------
+
+The JSON response is written by selectively applying ``json.dumps()``.
+The idea is to write some object layout manually, and leverage ``json.dumps()`` where possible.
+It basically looks like:
+
+.. code-block:: python
+
+    yield '{["_embedded": ['
+
+    # write main listing:
+    for record in generator:
+        if not first:
+            yield ",\n"
+        yield json.dumps(record)
+
+    yield "],\n"
+
+    # write "_links" and "page" parts without the opening/closing braces:
+    yield json.dumps(footer_links)[1:-1]
+    yield "}\n"
+
+
+Additional Optimizations
+------------------------
+
+Output Buffering
+~~~~~~~~~~~~~~~~
+
+To avoid too many back/forth calls from the response-generator
+and the WSGI server, the produced output is submitted in chunks of 4096kB.
+Otherwrite an OS ``write()`` call might happen for a simple ``yield "}"`` statement.
+
+Next Link Optimization
+~~~~~~~~~~~~~~~~~~~~~~
+
+A standard paginator would do an expensive ``COUNT(*)`` on the table
+and use that to tell whether thare are additional pages.
```

Comment:
and use that to tell whether _there_ are additional pages.

Reply:
This is digging into implementation details far too deeply. Just say that determining the number of pages is too expensive. That's true for the remotes as well.